### PR TITLE
removed atomic adds as far as possible

### PR DESCRIPTION
**Pull request by erikzenker**
* Improved cpu performance
* Still a lot of time is spend in importance sampling in the distribute rays randomly kernel

### CODE REVIEW

File: include/calc_sample_gain_sum.hpp [hunk 1]
```diff
@@ -268,59 +273,51 @@ struct CalcSampleGainSum {
...
-	    // the whole block gets a new offset (==workload)
-	    //rayNumber = getRayNumberBlockbased(acc, blockOffset, raysPerSample, globalOffsetMultiplicator);
-
-	    // HACK: ONLY WITH BLOCKSIZE 1, 1, 1
-	    blockOffset[0] = alpaka::atomic::atomicOp<alpaka::atomic::op::Add>(acc, globalOffsetMultiplicator, 1u);
-
-	    // rayNumber = alpaka::idx::getIdx<alpaka::Block, alpaka::Threads>(acc)[0] + (blockOffset[0] * alpaka::workdiv::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0]);
+	    alpaka::block::sync::syncBlockThreads(acc);
+	    // The first thread in the threadblock increases the globalOffsetMultiplicator (without real limit)
+	    if(alpaka::idx::getIdx<alpaka::Block, alpaka::Threads>(acc)[0] == 0){
```
Comment (slizzered) [inline include/calc_sample_gain_sum.hpp]: please decide if using tabs or spaces on this line

File: include/calc_sample_gain_sum.hpp [hunk 2]
```diff
@@ -268,59 +273,51 @@ struct CalcSampleGainSum {
...
-
-
+		
 		// Get triangle/prism to start ray from
-		unsigned startPrism             = indicesOfPrisms[rayNumber]; 
-		unsigned startLevel             = startPrism/mesh.numberOfTriangles;
-		unsigned startTriangle          = startPrism - (mesh.numberOfTriangles * startLevel);
-		Point startPoint                = mesh.genRndPoint(rand, startTriangle, startLevel);
-		Ray ray                         = generateRay(startPoint, samplePoint);
+	    	unsigned startPrism             = indicesOfPrisms[rayNumber];
```
Comment (slizzered) [inline include/calc_sample_gain_sum.hpp]: 4 lonely spaces managed to sneak in

File: include/importance_sampling.hpp [hunk 3]
```diff
@@ -60,22 +60,35 @@ struct DistributeRaysByImportance {
 				   const unsigned raysPerSample,
 				   unsigned *raysDump) const {
 
-	auto threadsVec = alpaka::workdiv::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc) *  alpaka::workdiv::getWorkDiv<alpaka::Block, alpaka::Threads>(acc);
-	auto nThreads = threadsVec[0];
+	auto threadsVec =
```
Comment (slizzered) [inline include/importance_sampling.hpp]: should be threadsExtent or something similar that reflects the 2-Dimensional nature of this variable :)

File: include/importance_sampling.hpp [hunk 4]
```diff
@@ -60,22 +60,35 @@ struct DistributeRaysByImportance {
...
+	unsigned tid              = alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0];
+	unsigned reflection_i     = alpaka::idx::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[1];
 	unsigned reflectionOffset = reflection_i * mesh.numberOfPrisms;
-
-	unsigned prism_i = alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0];
-	for(; prism_i < mesh.numberOfPrisms; prism_i += nThreads){	
-	    raysPerPrism[prism_i + reflectionOffset] = (unsigned) floor(importance[prism_i + reflectionOffset] / (*sumPhi) * raysPerSample);
+	unsigned raysDumpPart = 0;
+	
+	for(unsigned prism_i = tid; prism_i < mesh.numberOfPrisms; prism_i += nThreads){
```
Comment (slizzered) [inline include/importance_sampling.hpp]: Whitespace at end of Line :P

File: include/importance_sampling.hpp [hunk 5]
```diff
@@ -346,37 +352,33 @@ unsigned importanceSamplingDistribution(T_Stream &stream,
...
+    						      dRaysPerPrism,
+    						      raysPerSample,
+    						      alpaka::mem::view::getPtrNative(dRaysDump)));
+    	alpaka::stream::enqueue(stream, exec1);
+
+    	// alpaka::mem::view::copy(stream, hRaysDump, dRaysDump, static_cast<std::size_t>(1));
+    	// std::cout << "RaysLeft:" << raysPerSample - alpaka::mem::view::getPtrNative(hRaysDump)[0] << std::endl;
 	
-	alpaka::mem::view::getPtrNative(hRaysDump)[0] = raysPerSample;
+    	alpaka::mem::view::getPtrNative(hRaysDump)[0] = raysPerSample;
```
Comment (slizzered) [inline include/importance_sampling.hpp]: from here on, some spaces replaced the tabs

File: src/calc_phi_ase.cc [hunk 6]
```diff
@@ -275,7 +275,7 @@ float calcPhiAse ( const ExperimentParameters& experiment,
       
      	float hSumPhi = reduce(stream, dPreImportance, hMesh.numberOfPrisms * reflectionSlices, 0.);
 
-    	while(mseTooHigh){
+	while(mseTooHigh){
```
Comment (slizzered) [inline src/calc_phi_ase.cc]: yeah!

**Comment by erikzenker** [comment]
PUSH !

**Comment by slizzered** [comment]
very well done, thanks. I especially like the move towards a consistent whitespace use in our code :)
Commits are nice so far, I'll merge once you give the signal

**Comment by erikzenker** [comment]
I will run our test and give my ok if the result is correct.

**Comment by erikzenker** [comment]
Test crashed after 105 steps (infiniband cluster foo), but results look good!

| Alpaka |
|----------|
| 0.7992 |
| 0.8297 |
| 0.8607 |
| 0.8921 |
| 0.9240 |
| 0.9563 |
| 0.9889 |
| 1.0217 |
| 1.0548 |
| 1.0880 |
| 1.1212 |
| 1.1544 |
| 1.1876 |
| 1.2206 |
| 1.2534 |
| 1.2860 |
| 1.3182 |
| 1.3501 |
| 1.3815 |
| 1.4124 |
| 1.4428 |
| 1.4726 |
| 1.5017 |
| 1.5303 |
| 1.5581 |
| 1.5852 |
| 1.6115 |
| 1.6370 |
| 1.6618 |
| 1.6857 |
| 1.7089 |
| 1.7312 |
| 1.7528 |
| 1.7735 |
| 1.7934 |
| 1.8124 |
| 1.8306 |
| 1.8481 |
| 1.8648 |
| 1.8808 |
| 1.8960 |
| 1.9105 |
| 1.9244 |
| 1.9375 |
| 1.9500 |
| 1.9619 |
| 1.9731 |
| 1.9838 |
| 1.9939 |
| 2.0035 |
| 2.0125 |
| 1.9505 |
| 1.8935 |
| 1.8410 |
| 1.7924 |
| 1.7475 |
| 1.7057 |
| 1.6669 |
| 1.6306 |
| 1.5967 |
| 1.5649 |
| 1.5350 |
| 1.5069 |
| 1.4804 |
| 1.4554 |
| 1.4318 |
| 1.4094 |
| 1.3882 |
| 1.3681 |
| 1.3490 |
| 1.3308 |
| 1.3135 |
| 1.2970 |
| 1.2812 |
| 1.2662 |
| 1.2518 |
| 1.2381 |
| 1.2249 |
| 1.2123 |
| 1.2002 |
| 1.1886 |
| 1.1775 |
| 1.1668 |
| 1.1565 |
| 1.1466 |
| 1.1371 |
| 1.1279 |
| 1.1191 |
| 1.1105 |
| 1.1023 |
| 1.0944 |
| 1.0868 |
| 1.0794 |
| 1.0722 |
| 1.0653 |
| 1.0586 |
| 1.0522 |
| 1.0459 |
| 1.0399 |
| 1.0340 |

**Comment by slizzered** [comment]
good work finding those slow sections!

**Pull request merged**